Further improvements

NyakudyaA opened the pull request:
* Added the base build directory to the main repo
* Removed some duplications elsewhere
* Revert commit https://github.com/kartoza/docker-postgis/commit/e6bc0fb17dc80fa657fd58d35fdfa056bacf4ff5
* Fixes https://github.com/kartoza/docker-postgis/issues/278
* Upgrade container to use gdal3. Debian buster still install gdal 2.4 and PostGIS 13 is best served with gdal3
* Upgrade minor release to 13.1


PS: We can integrate GitHub workflows to build the base image

NyakudyaA committed:
add pg-cron back with the image

NyakudyaA committed:
Merge branch 'develop' of github.com:kartoza/docker-postgis into develop

NyakudyaA committed:
Merge branch 'develop' of github.com:kartoza/docker-postgis into develop

NyakudyaA committed:
added base build folder into main branch and various optimisations - revert commit for fixing passwords with strange characters - not worth the hassle

NyakudyaA committed:
split conf into seperate files

NyakudyaA committed:
DRY

NyakudyaA committed:
Change replication to streaming replication

NyakudyaA committed:
Fix build script to reflect building correct version

NyakudyaA committed:
make script executable

NyakudyaA committed:
Include extra config and fix wget in gdal install

NyakudyaA committed:
Some minor tweaks

NyakudyaA commented:
@lucernae I am not sure why the tests are failing but I tested on my local and all is working well

NyakudyaA commented:
ping @lucernae

lucernae commented:
I think I already fixed this before. Basically it’s just because the main image built are not built yet. Let me double check

lucernae commented:
Ok, it’s a different error. The base image failed to build, that’s why kartoza/postgis:manual-build is not ready yet.
There seems to be an issue in step:

```

E: Sub-process /usr/bin/dpkg returned an error code (1)
2608The command '/bin/sh -c set -eux     && export DEBIAN_FRONTEND=noninteractive     &&  apt-get update     && apt-get -y --no-install-recommends install postgresql-client-${POSTGRES_MAJOR_VERSION}         postgresql-common postgresql-${POSTGRES_MAJOR_VERSION}         postgresql-${POSTGRES_MAJOR_VERSION}-postgis-${POSTGIS_MAJOR}         netcat postgresql-${POSTGRES_MAJOR_VERSION}-ogr-fdw         postgresql-${POSTGRES_MAJOR_VERSION}-postgis-${POSTGIS_MAJOR}-scripts         postgresql-plpython3-${POSTGRES_MAJOR_VERSION} postgresql-${POSTGRES_MAJOR_VERSION}-pgrouting         postgresql-server-dev-${POSTGRES_MAJOR_VERSION} postgresql-${POSTGRES_MAJOR_VERSION}-cron' returned a non-zero code: 100

```

We should probably do `set -eux` in the build script so that it fails immediately when the build failed. Can’t tell much what happens in that commend. Probably missing env?

lucernae commented:
<strike>I got this from stderr:

```
W: GPG error: https://deb.meteo.guru/debian buster InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6BB3BABB751F80D7
E: The repository 'https://deb.meteo.guru/debian buster InRelease' is not signed.
```

This apt update right after using the base image. so we need to fix the base image.</strike>

Edit: sorry, it still picks up old branch caches. had to do a full build. Disregard above comments.

NyakudyaA commented:
> I got this from stderr:
> 
> ```
> W: GPG error: https://deb.meteo.guru/debian buster InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 6BB3BABB751F80D7
> E: The repository 'https://deb.meteo.guru/debian buster InRelease' is not signed.
> ```
> 
> This apt update right after using the base image. so we need to fix the base image.
> 
> Edit: sorry, it still picks up old branch caches. had to do a full build. Disregard above comments.

Roger this

lucernae committed:
Implement Github Action for base image

lucernae committed:
Modify GH Action for caching

lucernae committed:
Fix recipe and build matrix

lucernae committed:
Generate all locale only in deloyment mode

lucernae committed:
Fix tag name

lucernae committed:
Separate GH Action of build and deploy

lucernae committed:
Update locale generator logic

lucernae committed:
Set GH Action buildx to load images

lucernae committed:
Allow setting image owner

lucernae committed:
Test printing image list

lucernae committed:
Use multi-stage build

lucernae committed:
Optimize docker buildx cache

lucernae committed:
Add several default locale for testing

lucernae committed:
Fix docker-compose tty mode

lucernae committed:
Improve scenario tests logging

lucernae committed:
Improve build cache keys

lucernae committed:
Enable travis for comparison

lucernae commented:
@NyakudyaA Try pulling from https://github.com/kartoza/docker-postgis/tree/fix-ci
I've fixed some issues with the build, so now it builds using multi-stage.
Deployment at local can be done via `docker-compose build` with argument supplied from `.env` file.
Build test in CI was done via Github Action Workflows (like we wanted to do in the beginning of the PR), then we can replace Travis.

I added some custom build args too, so we probably need to update the docs. I will explain it in PR review after you do the pull from fix-ci to your branch.

<strike>I was not able to repair all the scenario tests, it seems the others are really broken.</strike>

Fixed now! There are some typos in the conf generator. So the tests indeed broke.

lucernae committed:
Fix scenario tests

- Fix cache keys for GH Action
- Disable logical_replication test health-check
  By disabling health-check, we can immediately see the logs

lucernae committed:
Fix travis

lucernae committed:
Goodbye Travis

I don't really know why the build failed in Travis, even though it is
using the same scripts.

Thank you for all the hardwork.

lucernae commented on .example.env:
```diff
@@ -0,0 +1,18 @@
+## Example environment file for docker-compose and builders
+
+COMPOSE_PROJECT_NAME=postgis
+
+## For build arguments
+DISTRO=debian
+IMAGE_VERSION=bullseye
+IMAGE_VARIANT=slim
+# Set GENERATE_ALL_LOCALE to empty value or 0 to build just default LOCALE: en_US.UTF-8
+GENERATE_ALL_LOCALE=1
+# Set the language if you need to specify LANG locale at build time
+LANG=en_US.UTF-8
+# locale filter to include in the locale generator
+LANGS="en_US.UTF-8,id_ID.UTF-8"
+
+POSTGRES_MAJOR_VERSION=13
+POSTGIS_MAJOR_VERSION=3
+POSTGIS_MINOR_RELEASE=1
```
Build arguments (and runtime docker-compose arguments) must be declared here.

Local user can do custom local build by copying this template into `.env` and fill in their values.
When we have changes to this file, user can diff their `.env` to this file for reference.

lucernae replied:
I extracted these:
```
POSTGRES_MAJOR_VERSION=13
POSTGIS_MAJOR_VERSION=3
POSTGIS_MINOR_RELEASE=1
```

From the Dockerfile recipe. Also included `POSTGIS_MAJOR_VERSION` to make it easier to understand.

lucernae commented on .example.env:
```diff
@@ -0,0 +1,18 @@
+IMAGE_VERSION=bullseye
+IMAGE_VARIANT=slim
+# Set GENERATE_ALL_LOCALE to empty value or 0 to build just default LOCALE: en_US.UTF-8
+GENERATE_ALL_LOCALE=1
+# Set the language if you need to specify LANG locale at build time
+LANG=en_US.UTF-8
+# locale filter to include in the locale generator
+LANGS="en_US.UTF-8,id_ID.UTF-8"
```
I added this build args to only build default locale.
Useful for creating slim base images to do the scenario tests (less time to build).
The reason why we use both en_US and id_ID is because it's used in the tests. So this is the default.

Both `LANG` and `LANGS` only used if `GENERATE_ALL_LOCALE=0` or empty (not generating all locale).

lucernae commented on .github/workflows/deploy-image.yaml:
```diff
@@ -0,0 +1,106 @@
+          context: .
+          file: Dockerfile
+          push: true
+          tags: ${{ secrets.DOCKERHUB_REPO }}/postgis:base-${{ matrix.distro }}-${{ matrix.imageVersion }}-${{ matrix.imageVariant }}
```
RFC to @NyakudyaA .

I propose we switch the image tag from `debian-bullseye-slim-base` into `base-debian-bullseye-slim`, due to the public convention that image variant is always the last, and `-` separates context from generic to specific (`base` being more generic usage).

Ideally there should be also the latest base tag which only spells `base`, like public convention.

lucernae commented on .github/workflows/deploy-image.yaml:
```diff
@@ -0,0 +1,106 @@
+          file: Dockerfile
+          push: true
+          load: true
+          tags: ${{ secrets.DOCKERHUB_REPO }}/postgis:${{ matrix.postgresMajorVersion }}-${{ matrix.postgisMajorVersion }}.${{ matrix.postgisMinorRelease }}
```
RFC to @NyakudyaA .

I propose we should also change the tag from `13.1` into `13-3.1` where the first part before dash is the postgres major version, and the rest is postgis full version to reduce ambiguity. 

Ideally we should also define the orchestration version somewhere in the back, like `13-3.1-21.01` where `21.01` is a shorthand for 2021 January. I prefer using `21.01` for orchestration tag (to version the Dockerfile recipe or build scripts) instead of semver because we don't yet have any clear convention to define a 'feature'.

The specific tag can be used for production usage (to really pin down the image), like `13-3.1-21.01.19`. But we should also track the common uses tag like `13` or `13.1` where user can easily type it for testing.

In practice, the common tag will always move. Today it can be that `13` is the same image as `13.1`, `13-3.1`, or `13-3.1-21.01` (in the order of specificity). Next month it could be that it points to `13-3.1-21-02`.

The one who will keep the tags to be up to date is the docker-meta actions (not yet being put here).

NyakudyaA replied:
I am ok with this. The convention was always that we always had one or two minor releases in each cycle. 

Something related to this versioning. Do we have a way to track if there is a new minor release so that we can automatically publish the corresponding image. Maybe we could investigate if we cannot do so for now

lucernae replied:
@NyakudyaA , did you mean tracking new minor release from upstream? (like from Postgis source repo?). If this is the case, we can only do that via dependency bot.

NyakudyaA replied:
Yes, But we can do it in another PR

lucernae commented on Dockerfile:
```diff
@@ -1,19 +1,85 @@
-#--------- Generic stuff all our Dockerfiles should start with so we get caching ------------
+##############################################################################
```
RFC to @NyakudyaA .

We are going to try to leverage Dockerfile Multi-stage build. I don't see any reasoning not too. It's more convenient, easier to maintain, and also cacheable. The only drawback that I regret is they can't be separate with some kind of `include` mechanism.

TL;DR; on how to leverage multi stage build:

We combine previous 3 Dockerfile (base, prod, and testing) into one but with 3 stages (base, prod, and testing).

Append extra parameters `--target` into the docker build (or docker-compose build) to specify which target you want to build.

Since we can refer the image inside the recipe (for example `postgis-base` is the name of the base image stage), caching happens without us having to push the image into registry. For example the  prod image is built using `FROM postgis-base AS postgis-prod`. We can do build like this:

```
# Build base image first (like we usually do)
docker build --target postgis-base .
# Build prod image, but it will start by using the cache from the postgis-base stage (no need to rebuild base)
docker build --target postgis-prod .
```

Or like this:
```
# Build prod image directly
docker build --target postgis-prod .
# Then build the base image as a tagged image, because we want to push the base over to registry
docker build -t kartoza/postgis:base --target postgis-base .
# In the above command, docker will use build cache because it already build postgis-base stage, when it builds postgis-prod
```

Basically we can keep the same usage workflow as usual, but with caching improvement in local and cloud build. The only thing we need to change is the build scripts.

lucernae commented on README.md:
```diff
@@ -59,8 +59,7 @@ docker pull kartoza/postgis:image_version
 
-To build the image yourself without apt-cacher (also consumes more bandwidth
-since deb packages need to be fetched each time you build) do:
+To build the image yourself do:
```
We need to update the docs, once this improvements are approved/finalized

lucernae commented on base_build/scripts/locale-filter.sh:
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env bash
+
+## Filter list of locales from a given filter args
+## Parse into array
+LANG_ARR=(${LANGS//,/ })
+echo "" > /etc/locale.gen
+for i in "${LANG_ARR[@]}"; do
+  cat /etc/all.locale.gen | grep "$i" >> /etc/locale.gen
+done
```
This is a helper script to filter desired locale in the build stage of postgis-base.
The scripts is unused if we build all locale

NyakudyaA replied:
Long overdue. Been a pending issue for people building locally a d Travis tests

lucernae commented on build-base.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
+# Building an debian base image
+
+if [[ ! -f .env ]]; then
+    echo "Default build arguments don't exists. Creating one from default value."
+    cp .example.env .env
+fi
+
+docker-compose -f docker-compose.build.yml build postgis-base
+
```
I moved this from `base_build` and adapt it to use docker-compose, multi-stage build and the `.env` file for build args

lucernae commented on build-test.sh:
```diff
@@ -1,5 +1,9 @@
 #!/usr/bin/env bash
+# For scenario testing purposes
 
-./build.sh
```
No need to specify base build. Because docker will understand the requirements from the stage declaration

lucernae commented on build-test.sh:
```diff
@@ -1,5 +1,9 @@
-./build.sh
+if [[ ! -f .env ]]; then
+    echo "Default build arguments don't exists. Creating one from default value."
+    cp .example.env .env
+fi
 
-docker build -t kartoza/postgis:manual-build -f Dockerfile.test .
+docker-compose -f docker-compose.build.yml build postgis-test
```
Simplify build command. Now the tags are declared in the docker-compose file. The stage are also declared in the docker-compose file. We only need to specify which service to build.

lucernae commented on build.sh:
```diff
@@ -1,3 +1,8 @@
 #!/usr/bin/env bash
-docker build -t kartoza/postgis:manual-build .
-docker build -t kartoza/postgis:13.0 .
```
Doesn't need to declare it. The dependencies are declared from the multi-stage build and declared in docker-compose file.

lucernae commented on docker-compose.build.yml:
```diff
@@ -0,0 +1,33 @@
+    image: kartoza/postgis:base-${DISTRO}-${IMAGE_VERSION}-${IMAGE_VARIANT}
+    build:
+      context: .
+      # Use yaml anchor for reusability
+      args: &build-args-anchor
+        DISTRO: ${DISTRO}
+        IMAGE_VERSION: ${IMAGE_VERSION}
+        IMAGE_VARIANT: ${IMAGE_VARIANT}
+        GENERATE_ALL_LOCALE: ${GENERATE_ALL_LOCALE}
+        POSTGRES_MAJOR_VERSION: ${POSTGRES_MAJOR_VERSION}
+        POSTGIS_MAJOR_VERSION: ${POSTGIS_MAJOR_VERSION}
+        POSTGIS_MINOR_RELEASE: ${POSTGIS_MINOR_RELEASE}
```
Using YAML anchor because all the build args used in all stages must be the same.

lucernae commented on docker-compose.build.yml:
```diff
@@ -0,0 +1,33 @@
+# Used solely for docker-compose build
```
This file is specific for building only. No runtime spec declared.

lucernae commented on scenario_tests/collations/test.sh:
```diff
@@ -8,17 +8,23 @@ source ../test-env.sh
 # Run service
 docker-compose up -d
 
+if [[ -n "${PRINT_TEST_LOGS}" ]]; then
+  docker-compose logs -f &
+fi
+
```
All the scenarios must have this so we can see the log in GH Action log as the tests runs asynchronously.

lucernae commented on scenario_tests/collations/test.sh:
```diff
@@ -8,17 +8,23 @@ source ../test-env.sh
 
   # Execute tests
-  until docker-compose exec $service pg_isready; do
+  until docker-compose exec -T $service pg_isready; do
```
All `docker-compose exec` must have `-T` flag so we can see the output in Github Action (it means no pseudo TTY allocation). But this is also harmless in local tests

lucernae commented on scenario_tests/logical_replication/docker-compose.yml:
```diff
@@ -47,8 +47,7 @@ services:
     depends_on:
-        pg-publisher:
-          condition: service_healthy
+        - pg-publisher
```
I had to explicitly do this. Because we don't want to wait for `pg-subscriber` to be up to show the logs. If something were happen to the service (causing `pg-publisher` not healthy), it will just wait block unless we declare it like this. So this is necessary for testing purposes.

NyakudyaA replied:
Maybe we could upgrade docker-compose from 2.1 to the latest version. The condition of service has been deprecated

lucernae commented on scripts/setup-conf.sh:
```diff
@@ -39,12 +38,14 @@ timezone='${TIMEZONE}'
+  if [[  -f ${ROOT_CONF}/logical_replication.conf ]];then
+    rm $CONF/logical_replication.conf
+  fi
+cat >> ${ROOT_CONF}/logical_replication.conf <<EOF
```
I fixed some typo. It was previously `streaming_replication.conf` so the test fails in `logical_replication` scenario

lucernae commented on .github/workflows/build-latest.yaml:
```diff
@@ -0,0 +1,197 @@
+name: build-latest
+on:
+#  workflow_dispatch:
+#    inputs:
+#      distro:
+#        description: Base image distro
+#        required: true
+#        default: debian
+#      imageVersion:
+#        description: Base distro image version/release
+#        required: true
+#        default: bullseye
+#      imageVariant:
+#        description: Base image variant
+#        required: true
+#        default: slim
```
This is commented for now, but can be useful if we have a use case to do an ad-hoc cloud build via GH Action.

For example, if the recipe can handle different base image. We can try to do a build from GH UI.

NyakudyaA replied:
Not much experience with actions

lucernae commented on .github/workflows/build-latest.yaml:
```diff
@@ -0,0 +1,197 @@
+        uses: actions/cache@v2
+        with:
+          path: /tmp/.buildx-cache
+          # Build inputs are:
+          # - Dockerfile recipe
+          # - docker-compose.build.yml
+          # - build args (.example.env)
+          # - base_build directory
+          # - scripts directory
+          key: buildx-${{ hashFiles('Dockerfile', 'docker-compose.build.yml', '.example.env', 'base_build', 'scripts') }}-${{ github.sha }}
+          restore-keys: |
+            buildx-${{ hashFiles('Dockerfile', 'docker-compose.build.yml', '.example.env', 'base_build', 'scripts') }}-
```
HashFiles of the build inputs are used to determine if we are doing the same build or not.
If it's the same, just retrieve from cache.

lucernae commented on .github/workflows/build-latest.yaml:
```diff
@@ -0,0 +1,197 @@
+            DISTRO=${{ matrix.distro }}
+            IMAGE_VERSION=${{ matrix.imageVersion }}
+            IMAGE_VARIANT=${{ matrix.imageVariant }}
+            LANGS=en_US.UTF-8,id_ID.UTF-8
+            GENERATE_ALL_LOCALE=0
```
For testing purposes, don't generate all locale.

lucernae reviewed (commented):
Ok. Here's my proposed changes:

Some changes (RFCs) needed some comment back because it will affect our convention in the future.
Thumbs up if you like the RFC. Otherwise we can discuss and make approriate changes in this PR.

lucernae commented:
@NyakudyaA , I can see that you agreed with the RFC.
Can we merge this so we can test action for build against all locale?

List of changes that can only be tested after this is merged in develop branch:
- Build against all locale
- Auto push to registry (merging this PR means we are trying to auto push `base-debian-bullseye-slim` and `13-3.1`
- Add short docker tag (like `base` and `13`)
- Do local build without push/login to docker registry on pull request (internal PR and external PR)
- Push test images on internal PR (PR from branches in this repo against develop)

Note that one of the workflow action above failing is because this is PR from external repo and we are using docker login inside. It failed because the Nyakudya:develop cannot access our secret (intentional for security). So, IMO, it's ok to ignore that for now. I'm going to fix that once the action is in develop.

NyakudyaA commented:
> @NyakudyaA , I can see that you agreed with the RFC.
> Can we merge this so we can test action for build against all locale?
> 
> List of changes that can only be tested after this is merged in develop branch:
> 
>     * Build against all locale
> 
>     * Auto push to registry (merging this PR means we are trying to auto push `base-debian-bullseye-slim` and `13-3.1`
> 
>     * Add short docker tag (like `base` and `13`)
> 
>     * Do local build without push/login to docker registry on pull request (internal PR and external PR)
> 
>     * Push test images on internal PR (PR from branches in this repo against develop)
> 
> 
> Note that one of the workflow action above failing is because this is PR from external repo and we are using docker login inside. It failed because the Nyakudya:develop cannot access our secret (intentional for security). So, IMO, it's ok to ignore that for now. I'm going to fix that once the action is in develop.

LGTM @lucernae You can proceed to merge. So in future if want to merge from my branch it will always fail because of the security for secrets. Is it safe to assume that if all the other tests are passing then the image is fine regardless of failure/success to push to dockerhub

lucernae commented:
> LGTM @lucernae You can proceed to merge. So in future if want to merge from my branch it will always fail because of the security for secrets. Is it safe to assume that if all the other tests are passing then the image is fine regardless of failure/success to push to dockerhub

We are going to do local build (inside github workflow) for external PR as it doesn't need docker login. (No need to setup secret for other user).
For internal PR, it's best that we push the PR image, so we can build the PR image. Something like kartoza/postgis:pr-123 then test it locally in our machine if we wish to. That means, checking PR from ourselves in our local machine doesn't need a local build. We just pull it from registry.

NyakudyaA commented:
> > LGTM @lucernae You can proceed to merge. So in future if want to merge from my branch it will always fail because of the security for secrets. Is it safe to assume that if all the other tests are passing then the image is fine regardless of failure/success to push to dockerhub
> 
> We are going to do local build (inside github workflow) for external PR as it doesn't need docker login. (No need to setup secret for other user).
> For internal PR, it's best that we push the PR image, so we can build the PR image. Something like kartoza/postgis:pr-123 then test it locally in our machine if we wish to. That means, checking PR from ourselves in our local machine doesn't need a local build. We just pull it from registry.

Great @lucernae, when this is merged I think I will also adapt it to the GeoServer repo

lucernae commented:
I will merge this now
